util/uniq/slice.go

```diff
@@ -6,60 +6,58 @@
 // It is similar to the unix command uniq.
 package uniq
 
-import (
-	"fmt"
-	"reflect"
-)
-
-type badTypeError struct {
-	typ reflect.Type
-}
-
-func (e badTypeError) Error() string {
-	return fmt.Sprintf("uniq.ModifySlice's first argument must have type *[]T, got %v", e.typ)
-}
+// ModifySlice removes adjacent duplicate elements from the given slice. It
+// adjusts the length of the slice appropriately and zeros the tail.
+//
+// ModifySlice does O(len(*slice)) operations.
+func ModifySlice[K comparable](slice *[]K) {
+	// Remove duplicates
+	dst := 0
+	for i := 1; i < len(*slice); i++ {
+		if (*slice)[i] == (*slice)[dst] {
+			continue
+		}
+		dst++
+		(*slice)[dst] = (*slice)[i]
+	}
 
-// ModifySlice removes adjacent duplicate elements from the slice pointed to by sliceptr.
-// It adjusts the length of the slice appropriately and zeros the tail.
-// eq reports whether (*sliceptr)[i] and (*sliceptr)[j] are equal.
-// ModifySlice does O(len(*sliceptr)) operations.
-func ModifySlice(sliceptr any, eq func(i, j int) bool) {
-	rvp := reflect.ValueOf(sliceptr)
-	if rvp.Type().Kind() != reflect.Ptr {
-		panic(badTypeError{rvp.Type()})
+	// Zero out the elements we removed at the end of the slice
+	end := dst + 1
+	var zero K
+	for i := end; i < len(*slice); i++ {
+		(*slice)[i] = zero
 	}
-	rv := rvp.Elem()
-	if rv.Type().Kind() != reflect.Slice {
-		panic(badTypeError{rvp.Type()})
+
+	// Truncate the slice
+	if end < len(*slice) {
+		*slice = (*slice)[:end]
 	}
+}
 
-	length := rv.Len()
+// ModifySliceFunc is the same as ModifySlice except that it allows using a
+// custom comparison function.
+//
+// eq should report whether (*slice)[i] and (*slice)[j] are equal.
+func ModifySliceFunc[K any](slice *[]K, eq func(i, j int) bool) {
```

Comment:
instead of taking `i, j int` you could make `eq func(a, b E)` or `eq func(a, b *E)` if you're worried about the size of parameter

Reply:
But now that I'm back at a computer, I don't see any ModifySlice in the tree that uses a custom eq func; only comparable values.

So adding this either way wouldn't break anybody at least. (or not adding it yet)

Reply:
Just removed this entirely; we can always re-add it later if necessary.